internal/npm_install/generate_build_file.js

```diff
@@ -0,0 +1,290 @@
+  return result;
+}
+
+/**
+ * Given the name of a top-level folder in node_modules, parse the
+ * package json and return it as an object along with
+ * some additional internal attributes prefixed with '_'.
+ */
+function parsePackage(p) {
+  // Parse the package.json file of this package
+  const pkg = JSON.parse(fs.readFileSync(`${p}/package.json`, {encoding: 'utf8'}));
+
+  // Trim the leading node_modules from the path and
+  // assign to _dir for future use
+  pkg._dir = p.replace(/^node_modules\//, '');
+
+  // Keep track of whether or not this is a nested package
+  pkg._isNested = p.match(/\/node_modules\//);
+
+  // Initialize _dependencies to an empty array
+  // which is later filled with the flattened dependency list
+  pkg._dependencies = [];
+
+  // For root packages, transform the pkg.bin entries
+  // into a new Map called _executables
+  pkg._executables = new Map();
+  if (!pkg._isNested) {
+    if (Array.isArray(pkg.bin)) {
+      // should not happen, but ignore it if present
+    } else if (typeof pkg.bin === 'string') {
+      pkg._executables.set(pkg._dir, cleanupBinPath(pkg.bin));
+    } else if (typeof pkg.bin === 'object') {
+      for (let key in pkg.bin) {
+        pkg._executables.set(key, cleanupBinPath(pkg.bin[key]));
+      }
+    }
+  }
+
+  return pkg;
+}
+
+/**
+ * Given a path, remove './' if it exists.
+ */
+function cleanupBinPath(path) {
+  // Bin paths usually come in 2 flavors: './bin/foo' or 'bin/foo',
+  // sometimes other stuff like 'lib/foo'.  Remove prefix './' if it
+  // exists.
+  path = path.replace(/\\/g, '/');
+  if (path.indexOf('./') === 0) {
+    path = path.slice(2);
+  }
+  return path;
+}
+
+/**
+ * Flattens all transitive dependencies of a package
+ * into a _dependencies array.
+ */
+function flattenDependencies(pkg, dep, pkgsMap) {
+  if (pkg._dependencies.indexOf(dep) !== -1) {
+    // circular dependency
+    return;
+  }
+  pkg._dependencies.push(dep);
+  const findDeps = function(targetDeps, required) {
+    Object.keys(targetDeps || {})
+        .map(targetDep => {
+          // look for matching nested package
+          const dirSegments = dep._dir.split('/');
+          while (dirSegments.length) {
+            const maybe = path.posix.join(...dirSegments, 'node_modules', targetDep);
+            if (pkgsMap.has(maybe)) {
+              return pkgsMap.get(maybe);
+            }
+            dirSegments.pop();
+          }
+          // look for matching root package
+          if (pkgsMap.has(targetDep)) {
+            return pkgsMap.get(targetDep);
+          }
+          // dependency not found
+          if (required) {
+            throw new Error(`Could not find required dep ${targetDep} of ${dep._dir}`)
+          }
+          return null;
+        })
+        .filter(dep => !!dep)
+        .map(dep => flattenDependencies(pkg, dep, pkgsMap));
+  };
+  findDeps(dep.dependencies, true);
+  findDeps(dep.peerDependencies, true);
+  findDeps(dep.optionalDependencies, false);
```

Comment:
I don't think the `optionalDependencies` thing works for solving #73. AFAIK both npm and yarn will attempt to install `optionalDependencies` but will continue if the install fails. The change to not error out of `optionalDependencies` were not found was needed for Windows actually, since there was an optional npm dependency on an npm package that did not install under Windows but it did under OSX/linux.

Reply:
The npm package that fails on Windows is `fsevents` which is an optionalDependency of `chokidar` in `build_bazel_rules_sass_compiletime_deps`. If `optionalDependencies` were required, you would get this error on Windows:
```
ERROR: C:/users/greg/_bazel_greg/34i6l3k3/external/io_bazel_rules_sass/sass/BUILD:11:1: no such package '@build_bazel_rules_sass_compiletime_deps//': ...
Error: Could not find required dep fsevents of chokidar
```
But with `fsevents` even as an optionalDependency of `chokidar`, it is still be installed on OSX/Linux, which is different than the semantics we would like for the `tsickle` case.

Reply:
I added extra comments here for clarification:
```
  // `optionalDependencies` that are missing should be silently
  // ignored since the npm/yarn will not fail if these dependencies
  // fail to install. Packages should handle the cases where these
  // dependencies are missing gracefully at runtime.
  // An example of this is the `chokidar` package which specifies
  // `fsevents` as an optionalDependency. On OSX/linux, `fsevents`
  // is installed successfully, but on Windows, `fsevents` fails
  // to install and the package will not be present when checking
  // the dependencies of `chokidar`.
  findDeps(dep.optionalDependencies, false);
```